### PR TITLE
feat(retry): bounded upstream retry/backoff for transient failures (#48)

### LINKED ISSUE #48: Bounded upstream retry/backoff for transient failures
## Summary

Add bounded, idempotent upstream retry/backoff for transient failures (connection resets, upstream `429`/`5xx`). This was planned in M4 but never implemented, so transient upstream errors currently surface directly to Claude Code.

## Why

A single transient upstream blip (TCP reset, brief `429`, upstream `502/503`) becomes a failed turn for the user, even though an immediate retry would likely succeed. A small, bounded retry with backoff and jitter improves resilience without adding scope or new dependencies. Must be careful about idempotency and streaming: only retry when nothing has been streamed to the client yet.

See `docs/comparison.md` §6 item E.

## Tasks

- [ ] Retry only pre-stream failures (before any SSE bytes reach the client) — never mid-stream.
- [ ] Bounded attempts (e.g. 2–3) with exponential backoff + jitter; honor `Retry-After` when present.
- [ ] Scope to clearly transient statuses (`429`, `502`, `503`, `504`, connection errors); never retry `4xx` request errors.
- [ ] Make attempts/backoff configurable per provider (default conservative); keep off for `count_tokens`.
- [ ] Tests for: transient → success on retry, non-transient → immediate surface, mid-stream → no retry.

## Evidence

- `docs/implementation-plan.md:247` (M4-planned retry/backoff, unimplemented)
- `src/reload.rs` is config hot-reload, not request retry (no retry path exists today)

## Related

- Comparison: `docs/comparison.md`

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

Adds a bounded, pre-stream, jittered exponential-backoff retry for transient upstream failures. On classified-transient errors (HTTP 429/502/503/504, or connection/timeout errors) the gateway retries obtaining the upstream `Response` before any body is streamed to the client, honoring `Retry-After` when present and giving up once the configured budget is exceeded. Because retry only operates on acquiring the response (not on data already streamed), it is structurally impossible to retry mid-stream. `count_tokens` requests are never retried.

## Milestone / spec

Closes #48 — bounded upstream retry/backoff for transient failures.

## Scope

Applies to single-credential paths only:
- Anthropic passthrough / `api_key`
- Responses API `api_key`, `xai_oauth`, and unpooled `chatgpt_oauth`
- Cursor

OAuth account pools (`claude_oauth`, `chatgpt_oauth` when pooled) are deliberately excluded — those paths already have account-rotation failover, so adding retry there would be redundant and could mask pool-health signals.

## Config

Retry is on by default with conservative settings:

```toml
[providers.<name>.retry]
max_retries = 2
initial_backoff_ms = 500
max_backoff_ms = 8000
multiplier = 2.0
```

Set `max_retries = 0` to disable retry for a provider. Config is validated at startup and on hot-reload (`max_retries <= 10`, `multiplier` finite and `>= 1.0`).

## Changes

- Per-provider bounded retry/backoff wrapper around upstream response acquisition, with jitter and `Retry-After` honoring
- Transient-failure classification (429/502/503/504 + connection/timeout errors) vs. immediate-surface for non-transient errors
- `[providers.<name>.retry]` config table + startup/hot-reload validation
- New `shunt.upstream_retries` metric, tagged by provider and reason
- Follow-up commit hardens classification, budget logging, and validation coverage per review feedback

## Test Plan

- [x] Transient failure followed by success is retried and succeeds
- [x] Non-transient failure surfaces immediately (no retry)
- [x] Budget-exhausted case returns the last response
- [x] Retry disabled (`max_retries = 0`) surfaces immediately
- [x] `count_tokens` requests are never retried
- [x] Successful first attempt is never retried (confirms no mid-stream retry path)
- [x] Responses-path retry covered
- [x] Cursor transient-failure classification covered
- [x] Config boundary/round-trip validation covered

## Checklist

- [x] `cargo build` passes
- [x] `cargo test` passes (new behavior is covered; tests run without network/loopback where possible)
- [x] `cargo clippy --all-targets --all-features -- -D warnings` clean
- [x] `cargo fmt --all --check` clean
- [x] Source files stay under 500 lines (new `src/retry.rs` is 484; `src/config.rs` was already >500 before this change)
- [x] English only; matches surrounding style
- [ ] Frozen spec in `docs/` updated if this change deviates from it (no spec deviation)
- [x] User-facing docs updated for behavior/config/endpoint/CLI/provider/model changes — `README.md`, `site/src/content/docs/reference/configuration.md`, `shunt.toml.example`
- [ ] Any new GitHub Action is pinned to a full commit SHA (no new Action)

## Notes for reviewers

- Retry only wraps acquisition of the upstream `Response`, before any body bytes are forwarded to the client — this is the mechanism that makes mid-stream retry structurally impossible, not just a policy choice.
- `cargo fmt --all --check`, `cargo clippy --all-targets --all-features -- -D warnings`, and the full `cargo test --all-features --workspace` suite were all run green on the final branch state (commit d8a048f).

### CODE REVIEW

File: src/retry.rs [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+        self.max_retries > 0
+    }
+}
+
+/// Transient upstream statuses that an identical, idempotent retry can plausibly
+/// clear. Deliberately excludes every other `4xx` (a request-level error) and
+/// `500` (which is frequently a deterministic upstream bug, not a blip).
+pub fn is_retryable_status(status: StatusCode) -> bool {
+    matches!(status.as_u16(), 429 | 502 | 503 | 504)
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/retry.rs:83]: @gemini-code-assist ✅ Applied — added `529` (Anthropic's non-standard "Overloaded") to `is_retryable_status`, so a transient overload from the Anthropic upstream is now retried at the gateway (still surfaced to the client on exhaustion, so Claude Code's own overload backoff is unaffected). Also updated the module doc and synced README / docs. Fixed in `0c66d83`.

File: src/retry.rs [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+/// A concise, low-cardinality reason label for the retry metric.
+fn retry_reason(status: StatusCode) -> &'static str {
+    match status.as_u16() {
+        429 => "429",
+        502 => "502",
+        503 => "503",
+        504 => "504",
+        _ => "other",
+    }
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/retry.rs:243]: @gemini-code-assist ✅ Applied — `retry_reason` now maps `529 => "529"` so the `shunt.upstream_retries` metric keeps 529 as a distinct low-cardinality label instead of collapsing to `"other"`. Covered by the `retry_reason_labels_each_transient_status_distinctly` test. `0c66d83`.

File: src/retry.rs [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+/// Decide how long to wait before the next attempt. An honored `Retry-After`
+/// wins over computed backoff; when it exceeds `max_backoff` the caller gives up
+/// cleanly instead of sleeping past budget.
+fn next_backoff(policy: &RetryPolicy, attempt: u32, headers: Option<&HeaderMap>) -> Backoff {
+    if let Some(retry_after) = headers.and_then(crate::accounts::retry_after) {
+        if retry_after > policy.max_backoff {
+            return Backoff::ExceedsBudget;
+        }
+        return Backoff::Sleep(retry_after);
+    }
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/retry.rs:254]: @greptileai ✅ Applied — `crate::accounts::retry_after` now parses both RFC 7231 forms: delta-seconds first, then the HTTP-date form (via the `httpdate` crate, already transitive in the tree, promoted to a direct dep). A past date resolves to "retry now" (zero wait) rather than falling through to computed backoff. Added tests for the future-date, past-date, and unparseable cases. Same issue cubic flagged. `0c66d83`.

File: src/adapters/anthropic/mod.rs [hunk 4]
```diff
@@ -53,14 +53,29 @@ async fn forward(
...
+    // `Bytes` clones the body as a cheap refcount bump, so an idempotent re-send
+    // on retry never re-copies it.
+    let body = bytes::Bytes::from(body);
+    let client = state.http_client.clone();
+    let upstream = crate::retry::send_with_retry(policy, &route.provider, || {
+        client
+            .post(url.as_str())
+            .headers(request_headers.clone())
+            .body(body.clone())
+            .send()
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/adapters/anthropic/mod.rs:75]: @cubic-dev-ai ⛔ Valid concern, but declining the suggested fix on this PR — tracked as #126 instead.

The shared reqwest client configures **no response/read timeout** (`src/main.rs`), so `reqwest::Error::is_timeout()` only fires connect-phase — *before* Anthropic accepts the POST. A post-acceptance response timeout cannot occur today. The genuine residual case is retrying a `502`/`504` whose origin already processed the request, and the suggested remedy can't eliminate it: the Anthropic **Messages** API exposes **no idempotency key**. Retrying transient `5xx` without one matches the Anthropic SDK's own default. The risk is bounded (strictly pre-stream, capped `max_retries`) and now documented in the `src/retry.rs` module doc. Hardening is tracked in #126.

File: src/retry.rs [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+    mut attempt: F,
+) -> Result<reqwest::Response, E>
+where
+    E: RetryableError + std::fmt::Display,
+    F: FnMut() -> Fut,
+    Fut: std::future::Future<Output = Result<reqwest::Response, E>>,
+{
+    let mut retries: u32 = 0;
+    loop {
+        let outcome = attempt().await;
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/retry.rs:134]: @cubic-dev-ai ⛔ Declining the fix on this PR; hardening tracked in #126.

This loop retries only (a) transient `5xx`/`429` **responses** — where the upstream received and rejected the request — or (b) connect-phase transport errors. Because the client configures no response/read timeout, `is_timeout()` is pre-acceptance. The only residual duplicate risk is a `502`/`504` the origin already processed, which no idempotency key can dedupe here — Messages / Responses / Cursor expose none. Bounded + documented in the module doc. See #126.

File: src/adapters/cursor/mod.rs [hunk 6]
```diff
@@ -112,10 +112,22 @@ async fn forward(
...
+        .map(|provider| provider.retry.policy())
+        .unwrap_or(crate::retry::RetryPolicy::DISABLED);
+    let client = CursorHttpClient::new(state.http_client.clone(), base_url);
+    // Cursor has no account-pool failover, so a transient 429/5xx or connection
+    // blip on the single credential would otherwise surface straight to the
+    // client. Bounded retry (issue #48) stays pre-stream: it only re-runs
+    // `run_agent` — which mints a fresh request id per attempt — before the
+    // `is_success` check below hands off to the streaming/bridge decoders.
+    let upstream = crate::retry::send_with_retry(policy, &route.provider, || {
+        client.run_agent(&access_token, &prompt, &resolved, &images)
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/adapters/cursor/mod.rs:127]: @cubic-dev-ai ⛔ Declining on this PR; captured in #126.

Cursor `Run` retries only on connect-phase failures (pre-acceptance — no response timeout is configured) or transient statuses where Cursor rejected the request without starting a run, so a fresh request id per attempt does not create a duplicate run in practice. A true post-acceptance timeout can't occur today. Threading one idempotency identity across attempts (incl. `x-original-request-id`) is a good hardening — tracked in #126.

File: src/adapters/responses/http.rs [hunk 7]
```diff
@@ -68,10 +77,25 @@ pub(super) async fn forward_http(
...
-    // Send via the shared helper (extracted so the account-pool path can classify
-    // a response before relaying); it wraps the same request_builder + send used
-    // above the merge with #112's estimate overlap.
-    let upstream = http_send(state, route, credential, session_id, &upstream_body).await?;
+    // The account-pool path drives its own failover and deliberately does not
+    // layer retry on top. This single-credential path retries only before any
+    // response body is handed to the streaming/JSON relay.
+    let policy = provider_retry_policy(state, route);
+    let body = bytes::Bytes::from(upstream_body.to_string());
+    let upstream = crate::retry::send_with_retry(policy, &route.provider, || {
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/adapters/responses/http.rs:85]: @cubic-dev-ai ⛔ Declining on this PR; tracked in #126.

Same reasoning as the Anthropic path: the shared client has no response/read timeout, so `is_timeout()` is connect-phase (pre-acceptance); the residual `502`/`504`-already-processed case has no idempotency key on the Responses `/responses` endpoint to dedupe against. Bounded (pre-stream, capped retries) and documented in the `retry.rs` module doc. Hardening in #126.

File: src/adapters/responses/http.rs [hunk 8]
```diff
@@ -68,10 +77,25 @@ pub(super) async fn forward_http(
...
+    .map_err(|error| {
+        // Preserve the raw transport cause in logs before own_error maps it to
+        // the stable gateway-facing Responses error envelope.
+        tracing::warn!(
+            provider = %route.provider,
+            error = %error,
+            "responses upstream request failed after retries"
+        );
+        own_error(error.to_string())
+    })?;
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/adapters/responses/http.rs:98]: @gemini-code-assist 🚫 The specific transport error isn't actually lost here. At this layer it's already logged with full context — `tracing::warn!(provider, error = %error, "responses upstream request failed after retries")`, i.e. the same `%error` form (not `%error.message`) that the pool.rs change you cite standardized on — and the concrete message flows into the **client-facing response body** via `own_error` → `ShuntError::bad_gateway(message)`. Only the upper-layer `AdapterError.message` label is generic; duplicating the raw error there would just re-log what's already captured at the point of failure. The suggested `UpstreamError::from_reqwest` would also swap the deliberate gateway-owned Responses error envelope (`own_error`) for the Anthropic-shaped one, which this path intentionally doesn't use. Leaving as-is.

File: src/retry.rs [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+pub struct RetryPolicy {
+    /// Additional attempts after the first. `0` disables retry entirely.
+    pub max_retries: u32,
+    /// Backoff ceiling for the first retry, grown by `multiplier` per attempt.
+    pub initial_backoff: Duration,
+    /// Upper bound on any single backoff and on an honored `Retry-After`.
+    pub max_backoff: Duration,
+    /// Exponential growth factor applied per attempt (>= 1.0).
+    pub multiplier: f64,
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/retry.rs:59]: @gemini-code-assist ⛔ The invariants you list are already enforced at the **sole production construction site**. `RetryConfig::validate` (`src/config.rs`) rejects `max_retries > 10`, a non-finite or `< 1.0` multiplier, and zero backoffs when `max_retries > 0`; `RetryConfig::policy()` is the only place a runtime `RetryPolicy` is built (plus the hardcoded `DISABLED`). `RetryPolicy` is an internal `Copy` DTO threaded from that already-validated config to the call site — making its fields private behind a fallible `TryFrom` would duplicate the config-layer validation and force test-only construction escape hatches for the driver's edge-case tests, with no real safety gain (YAGNI). Declining; the type-level guarantee already lives at the config boundary.
